content/blog/2020-04-preemption.md

```diff
@@ -0,0 +1,207 @@
+definition of progress is whether or not a task has been scheduled for over some
+unit of time. For example, if a worker has been stuck scheduling the same task
+for more than 100ms, then that worker is flagged as blocked and a new thread is
+spawned. In this definition, how does one detect scenarios where spawning a new
+thread **reduces** throughput? This can happen when the scheduler is generally
+under load and adding threads would make the situation much worse. To combat
+this, the .NET thread pool uses [hill climbing][hill]. [This article][hill2]
+provides a good overview of how it works.
+
+The second problem is that any automatic detection strategy will be vulnerable
+to bursty or otherwise uneven workloads. This specific problem has been the bane
+of the .NET thread pool and is known as the ["stuttering" problem][stutter]. The
+hill climbing strategy requires some period of time (hundreds of milliseconds)
+to adapt to load changes. This time period is needed, in part, to be able to
+determine that adding threads is improving the situation and not making it
+worse.
+
+The stuttering problem can be managed with the .NET thread pool, in part,
+because the pool is designed to schedule **coarse** tasks, i.e. tasks that
+execute in the order of hundreds of milliseconds to multiple seconds. However,
+in Rust, asynchronous task schedulers are designed to schedule tasks that should run in
+the order of microseconds to tens of milliseconds at most. In this case, any
+stutttering problem from a heuristic-based scheduler will result in far greater
+latency variations.
+
+The most common follow-up question I've received after this is "doesn't the Go
+scheduler automatically detect blocked tasks?". The short answer is: no. Doing
+so would result in the same stuttering problems as mentioned above. Also, Go has
+no need to have generalized blocked task detection because Go is able to
+preempt. What the Go scheduler **does** do is annotate potentially blocking
+system calls. This is roughly equivalent to the Tokio's
+[`block_in_place`][block_in_place].
+
+In short, as of now, the automatic cooperative task yielding strategy that has
+just been introduced is the best we have found for reducing tail latencies.
+Because this strategy only requires Tokio's types to opt-in, the end user does
+not need to change anything to gain this benefit. Simply upgrading the Tokio
+version will include this new functionality. Also, if Tokio's types are used
+from **outside** of the Tokio runtime, they will behave as they did before.
+
+There is more work that should happen on this topic. It is still how unclear how
+task budgets should work with "sub schedulers" (e.g.
+[`FuturesUnordered`][futunord]). The task budget APIs should eventually be
+exposed so third party libs can integrate with them. It also would be nice to
+figure out a way to generalize this concept so more than just Tokio users can
+take advantage of this.
```

Comment:
Maybe the "future work" section could end with a "call to action"? "here's how you can help out if you're interested..." or even just "try it out by updating to today's tokio release"?